### PR TITLE
Update libcudf counting functions to specify cudf::size_type

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Adds section to developer guide about `cudf::size_type` and adds links to it from other relevant parts of the document.
The fundamental nature of this type seems important enough to mention in the developer guide since it is the basis for how much of the code is designed and implemented.
Also updates some doxygen for public APIs that are return `size_type` column values but had cited `INT32` specifically.

Reference: https://github.com/rapidsai/cudf/pull/12779#discussion_r1117315749

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 1]
```diff
@@ -747,7 +754,7 @@ The "Indexalator", or index-normalizing iterator (`include/cudf/detail/indexalat
 used for index types (integers) without requiring a type-specific instance. It can be used for any
 iterator interface for reading an array of integer values of type `int8`, `int16`, `int32`,
 `int64`, `uint8`, `uint16`, `uint32`, or `uint64`. Reading specific elements always return a
```
Comment (bdice) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:756]:
```suggestion
`int64`, `uint8`, `uint16`, `uint32`, or `uint64`. Reading specific elements always returns a
```

File: cpp/include/cudf/strings/attributes.hpp [hunk 2]
```diff
@@ -32,39 +32,39 @@ namespace strings {
...
  * The output column will have the same number of rows as the
  * specified strings column. Each row value will be the number of
  * bytes in the corresponding string.
  *
  * Any null string will result in a null entry for that row in the output column.
  *
- * @param strings Strings instance for this operation.
- * @param mr Device memory resource used to allocate the returned column's device memory.
- * @return New INT32 column with the number of bytes for each string.
+ * @param input Strings instance for this operation
```
Comment (bdice) [inline cpp/include/cudf/strings/attributes.hpp:62]: We deleted the periods on this docstring but left the periods on some other ones. I don't remember our preferred convention.

File: cpp/include/nvtext/tokenize.hpp [hunk 3]
```diff
@@ -116,7 +116,7 @@ std::unique_ptr<cudf::column> tokenize(
  * @param delimiter Strings used to separate each string into tokens.
  *                  The default of empty string will separate tokens using whitespace.
  * @param mr Device memory resource used to allocate the returned column's device memory.
- * @return New INT32 column of token counts.
+ * @return New column of cudf::size_type of token counts
```
Comment (bdice) [inline cpp/include/nvtext/tokenize.hpp:119]: Some of the return type changes above removed information about the type entirely (e.g. deleting INT32). Does this function need to specially note its data type?

**Review by bdice — COMMENTED** [review]
Just a few minor comments. This is a significant improvement in clarity for our library design. Thank you very much!

File: cpp/include/cudf/strings/attributes.hpp (hunk 2 above)
Comment (ttnghia) [reply cpp/include/cudf/strings/attributes.hpp:62]: As I noticed, people now almost always eliminate periods.

Comment (davidwendt) [reply cpp/include/cudf/strings/attributes.hpp:62]: The convention is to use periods only on complete sentences.

File: cpp/include/nvtext/tokenize.hpp (hunk 3 above)
Comment (davidwendt) [reply cpp/include/nvtext/tokenize.hpp:119]: Considering the awkward wording, I'm starting to prefer not including the data type in the `@return`.

**Review by karthikeyann — COMMENTED** [review]
At some `@return` doc, `size_type` is removed, but left `cudf::size_type` in some `@brief`. 
What is the convention followed here?

File: cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md [hunk 4]
```diff
@@ -1224,7 +1231,7 @@ libcudf provides view types for nested column types as well as for the data elem
 `cudf::strings_column_view` is a view of a strings column, like `cudf::column_view` is a view of
 any `cudf::column`. `cudf::string_view` is a view of a single string, and therefore
 `cudf::string_view` is the data type of a `cudf::column` of type `STRING` just like `int32_t` is the
-data type for a `cudf::column` of type `INT32`. As it's name implies, this is a read-only object
+data type for a `cudf::column` of type [`size_type`](#cudfsize_type). As it's name implies, this is a read-only object
```
Comment (wence-) [inline cpp/doxygen/developer_guide/DEVELOPER_GUIDE.md:1234]:
```suggestion
data type for a `cudf::column` of type [`size_type`](#cudfsize_type). As its name implies, this is a read-only object
```

**Review by wence- — COMMENTED** [review]
One grammar change (while we're changing the line anyway)

**Comment by davidwendt** [comment]
I believe all comments have been addressed now.

**Review by wence- — APPROVED** [review]
My approval carries no official weight here I think, but looks good, thanks!

**Review by PointKernel — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**